File: newSolver.py
import math
import time
from enum import Enum
from testData import *

class Status(Enum):
    SOLVED = 0
    IN_PROGRESS = 1
    IMPOSSIBLE = 2

class Grid(object):
    def __init__(self, size, secW, secH, depth=0):
        if (size % secW + size % secH) != 0:
            return
        
        self._size = size
        self._secW = secW # Width of each section
        self._secH = secH # Height of each section
        self._nSecsW = math.floor(size/self._secW) # Number of sections along width
        self._nSecsH = math.floor(size/self._secH) # Number of sections along height
        self._status = Status.IN_PROGRESS
        self._solved = []
        self._possible = []
        self._nSolved = 0
        self._queue = []
        self._depth = depth

        for i in range(size):
            solvedRow = []
            posRow = []
            for j in range(size):
                solvedRow.append(False)
                posRow.append(self.allNumbers())
            self._solved.append(solvedRow)
            self._possible.append(posRow)

    def load(self,sampleGrid):
        if len(sampleGrid) != self._size:
            return False
        for row in sampleGrid:
            if len(row) != self._size:
                return False
        
        # nSolved = 0
        for i in range(self._size):
            for j in range(self._size):
                n = sampleGrid[i][j]
                if (n != 0):
                    self.addToQueue(n,i,j)

    def allNumbers(self):
        numbers = []
        for i in range(self._size):
            numbers.append(1+i)
        return numbers

    def printOut(self,showPossible=False):
        display = []
        for i in range(0,self._size):
            tempRow = []
            for j in range(0,self._size):
                # print(self._solved[i][j])
                if (self._solved[i][j]):
                    tempRow += [self._possible[i][j][0]]
                else:
                    tempRow += ["?"]
            display += [tempRow]

        for i in range(0, self._size):
            tempRow = ""
            for j in range(0,self._size-1):
                tempRow += str(display[i][j]) + " "
                if ((j + 1) % self._secW) == 0:
                    tempRow += "# "
            tempRow += str(display[i][self._size-1])
            print(tempRow)

            if ((i + 1) % self._secH) == 0 and i != self._size - 1:
                tempRow = "# "*(self._size + round(self._size/self._secW)-2) + "#"
                print(tempRow)

        if (showPossible):
            for i in range(self._size):
                for j in range(self._size):
                    if (not self._solved[i][j]):
                        print(i,j,self._possible[i][j])

    def writeIn(self,n,x,y):
        # print("Writing in",x,y,"at",n)
        # if (self._status != Status.IN_PROGRESS):
        #     print("Dying here")
        #     return False

        if (self._solved[x][y]):
            if (self._possible[x][y][0] != n):
                self._status = Status.IMPOSSIBLE
                self._nSolved = 0
                return False
            else:
                return True

        self._solved[x][y] = True
        self._possible[x][y] = [n]
        self._nSolved += 1

        # print(self._depth*"-",x,y,":",n)

        if self._nSolved == self._size**2:
            self._status = Status.SOLVED
            return True

        for j in range(self._size):
            if (j == y):
                continue
            if (n in self._possible[x][j]):
                self._possible[x][j].remove(n)
                if (len(self._possible[x][j]) == 0):
                    self._status = Status.IMPOSSIBLE
                    self._nSolved = 0
                    # print("Reached an impossible place")
                    return False
                elif (len(self._possible[x][j]) == 1 and not self._solved[x][j]):
                    self.addToQueue(self._possible[x][j][0],x,j)

        for i in range(self._size):
            if (i == x):
                continue
            if (n in self._possible[i][y]):
                self._possible[i][y].remove(n)
                if (len(self._possible[i][y]) == 0):
                    self._status = Status.IMPOSSIBLE
                    self._nSolved = 0
                    # print("Reached an impossible place")
                    return False
                elif (len(self._possible[i][y]) == 1 and not self._solved[i][y]):
                    self.addToQueue(self._possible[i][y][0],i,y)
        
        tl = self.getTLofSec(x,y)
        for i in range(self._secH):
            for j in range(self._secW):
                if (tl[0]+i == x and tl[1]+j == y):
                    continue
                if (n in self._possible[tl[0]+i][tl[1]+j]):
                    self._possible[tl[0]+i][tl[1]+j].remove(n)
                    if (len(self._possible[tl[0]+i][tl[1]+j]) == 0):
                        self._status = Status.IMPOSSIBLE
                        # print("Reached an impossible place")
                        return False
                    elif (len(self._possible[tl[0]+i][tl[1]+j]) == 1 and not self._solved[tl[0]+i][tl[1]+j]):
                        self.addToQueue(self._possible[tl[0]+i][tl[1]+j][0],tl[0]+i,tl[1]+j)
                
        return True

    def getTLofSec(self,x,y):
        tl = [0,0]
        tl[0] = self._secH*math.floor(x/self._secH)
        tl[1] = self._secW*math.floor(y/self._secW)
        return tl

    def addToQueue(self,n,x,y):
        self._queue.append([n,x,y])

    def solveByElimination(self):
        if (self._status != Status.IN_PROGRESS):
            return False
        # Clear each row
        # Clear each column
        # Clear each block
        currentSolved = self._nSolved
        for i in range(self._size):
            row = []
            for j in range(self._size):
                row.append([i,j])
            self.solveGroup(row)

        for j in range(self._size):
            col = []
            for i in range(self._size):
                row.append([i,j])
            self.solveGroup(col)

        for x in range(self._nSecsH):
            for y in range(self._nSecsW):
                tlX = self._secH*x
                tlY = self._secW*y
                sec = []
                for i in range(self._secH):
                    for j in range(self._secW):
                        sec.append([tlX+i,tlY+j])
                self.solveGroup(sec)

        return (self._nSolved > currentSolved)

    
    def solveGroup(self,group):
        numbers = self.allNumbers()
        emptys = []
        # Find out what numbers aren't accounted for
        # Find out what available spaces there are
        for pos in group:
            if (self._solved[pos[0]][pos[1]]):
                # print(pos[0],pos[1],self._possible[pos[0]][pos[1]])
                try:
                    numbers.remove(self._possible[pos[0]][pos[1]][0])
                except:
                    self._status = Status.IMPOSSIBLE
                    # print("Reached an impossible state")
                    return
            else:
                emptys.append(pos)
        
        for n in numbers:
            chosen = None
            found = False
            for pos in emptys:
                # May be solved in the process
                if self._solved[pos[0]][pos[1]]:
                    continue
                if n in self._possible[pos[0]][pos[1]]:
                    if found:
                        found = False
                        break
                    else:
                        chosen = pos
                        found = True
            if found:
                self.writeIn(n,chosen[0],chosen[1])
                self.doQueue()
        
    def doQueue(self):
        currentSolved = self._nSolved
        while (len(self._queue) != 0):
            self.writeIn(self._queue[0][0],self._queue[0][1],self._queue[0][2])
            self._queue = self._queue[1:]
        return (self._nSolved > currentSolved)

    def clone(self):
        newGrid = Grid(self._size,self._secW,self._secH,self._depth+1)
        for i in range(0,self._size):
            for j in range(0,self._size):
                if (self._solved[i][j]):
                    newGrid.writeIn(self._possible[i][j][0],i,j)
        # print("Clone:")
        # newGrid.printOut()
        return newGrid

    def getNatXY(self,x,y):
        if (self._solved[x][y]):
            assert len(self._possible[x][y]) == 1
            return self._possible[x][y][0]
        else:
            return 0

    def getStatus(self):
        return self._status

    def adapt(self,otherGrid):
        for i in range(0,self._size):
            for j in range(0,self._size):
                self.writeIn(otherGrid.getNatXY(i,j),i,j)
        self._status = otherGrid.getStatus()
        # for i in range(0,self._size):
        #     for j in range(0,self._size):
        #         print(self.getNatXY(i,j))

    def solveByGuess(self):
        if (self._status != Status.IN_PROGRESS):
            return False

        changed = False
        for i in range(0,self._size):
            for j in range(0,self._size):
                if (not self._solved[i][j] and len(self._possible[i][j]) <= 4):
                    works = -1
                    nWorks = 0
                    k = 0
                    while (k < len(self._possible[i][j])):
                        # print(k,len(self._possible[i][j]))
                        tempGrid = self.clone()
                        tempGrid.writeIn(self._possible[i][j][k],i,j)
                        # print("depth:",self._depth,'  '*self._depth,"Guessing",self._possible[i][j][k],"at",i,j)
                        tempGrid.solve()
                        if (tempGrid.getStatus() == Status.SOLVED):
                            # print("Hey this one worked")
                            nWorks = 1
                            self.adapt(tempGrid)
                            return True
                        # elif (tempGrid.getStatus() == Status.IN_PROGRESS):
                        if (tempGrid.getStatus() != Status.IMPOSSIBLE):
                            nWorks += 1
                            works = self._possible[i][j][k]
                            if (nWorks >= 2):
                                break
                        k += 1
                    if (nWorks == 1):
                        # print("Found that",works,"Works")
                        changed = True
                        self.writeIn(works,i,j)
                        # print("Depth:",self._depth,'  '*self._depth,"GOOD:",works,"at",i,j)
                        progress = True
                        while progress:
                            progress = (self.doQueue() or self.solveByElimination())
        return changed

    def solve(self):
        progress = True
        while progress:
            # print("going")
            progress = (self.doQueue() or self.solveByElimination())
        if (self._status == Status.IN_PROGRESS):
            print("Done all logical steps, moving onto guessing")
            if (self._depth < 2):
                self.solveByGuess()

    def getGrid(self):
        ret = []
        for i in range(self._size):
            row = []
            for j in range(self._size):
                if self._solved[i][j]:
                    row.append(self._possible[i][j][0])
                else:
                    row.append(0)

            ret.append(row)
        return ret
        

if __name__ == "__main__":
    mainGrid = Grid(9,3,3)
    # mainGrid.writeIn(1,5,5)
    # nSolved = 0
    # for i in range(9):
    #     for j in range(9):
    #         n = exp2Numbers[i][j]
    #         if (n != 0):
    #             mainGrid.addToQueue(n,i,j)
    #             nSolved += 1
    mainGrid.load(evilNumbers)
    start = time.process_time()
    mainGrid.solve()
    print("TIME:",time.process_time() - start, mainGrid.getStatus())
    mainGrid.printOut()
    grid = mainGrid.getGrid()
    print(grid)
    # print("nSolved:",nSolved,'->',mainGrid._nSolved)

File: docs/newSolver.js
const Status = {
    SOLVED: 'solved',
    IN_PROGRESS: 'in progress',
    IMPOSSIBLE: 'impossible'
}


class Grid {
    constructor(size,secW,secH,depth=0){
        this.size = size
        this.secW = secW
        this.secH = secH
        this.nSecsW = Math.floor(size/secW)
        this.nSecsH = Math.floor(size/secH)
        this.status = Status.IN_PROGRESS
        this.solved = []
        this.possible = []
        this.nSolved = 0
        this.queue = []
        this.depth = depth

        for (let i = 0; i < size; i++){
            let solvedRow = []
            let posRow = []
            for (let j = 0; j < size; j++){
                solvedRow.push(false)
                posRow.push(this.allNumbers())
            }
            this.solved.push(solvedRow)
            this.possible.push(posRow)
        }
    }

    async load(sampleGrid){
        if (sampleGrid.length != this.size){
            return false
        }
        for (let row of sampleGrid){
            if (row.length != this.size){
                return false
            }
        }

        for (let i = 0; i < this.size; i++){
            for (let j = 0; j < this.size; j++){
                let n = sampleGrid[i][j]
                if (n != 0){
                    this.addToQueue(n,i,j)
                }
            }
        }

        return true
    }

    allNumbers() {
        return [...Array(this.size).keys()].map(x => x + 1)
    }

    async writeIn(n,x,y){
        if (this.solved[x][y]){
            if (this.possible[x][y][0] != n){
                this.status = Status.IMPOSSIBLE
                this.nSolved = 0
                return false
            } else {
                return true
            }
        }

        this.solved[x][y] = true
        this.possible[x][y] = [n]
        this.nSolved++

        // console.log([this.depth,x,y,n])

        if (this.nSolved == this.size*this.size){
            this.status = Status.SOLVED
            return true
        }

        for (let j = 0; j < this.size; j++){
            if (j == y){
                continue
            }
            let index = this.possible[x][j].indexOf(n)
            if (index != -1){
                this.possible[x][j].splice(index,1)
                if (this.possible[x][j].length == 0){
                    this.status = Status.IMPOSSIBLE
                    this.nSolved = 0
                    return false
                } else if ((this.possible[x][j].length == 1) && !(this.solved[x][j])){
                    this.addToQueue(this.possible[x][j][0],x,j)
                }
            }
        }

        for (let i = 0; i < this.size; i++){
            if (i == x){
                continue
            }
            let index = this.possible[i][y].indexOf(n)
            if (index != -1){
                this.possible[i][y].splice(index,1)
                if (this.possible[i][y].length == 0){
                    this.status = Status.IMPOSSIBLE
                    this.nSolved = 0
                    return false
                } else if ((this.possible[i][y].length == 1) && !(this.solved[i][y])){
                    this.addToQueue(this.possible[i][y][0],i,y)
                }
            }
        }

        let tl = this.getTLofSec(x,y)
        for (let i = 0; i < this.secH; i++){
            for (let j = 0; j < this.secW; j++){
                if ((tl[0]+i == x) && (tl[1]+j == y)){
                    continue
                }
                let index = this.possible[tl[0]+i][tl[1]+j].indexOf(n)
                if (index != -1){
                    this.possible[tl[0]+i][tl[1]+j].splice(index,1)
                    if (this.possible[tl[0]+i][tl[1]+j].length == 0){
                        this.status = Status.IMPOSSIBLE
                        this.nSolved = 0
                        return false
                    } else if ((this.possible[tl[0]+i][tl[1]+j].length == 1) && !(this.solved[tl[0]+i][tl[1]+j])){
                        this.addToQueue(this.possible[tl[0]+i][tl[1]+j][0],tl[0]+i,tl[1]+j)
                    }
                }
            }
        }

        return true
    }

    getTLofSec(x,y){
        let i = this.secH * Math.floor(x/this.secH)
        let j = this.secW * Math.floor(y/this.secW)
        return [i,j]
    }

    addToQueue(n,x,y){
        this.queue.push([n,x,y])
    }

    async solveByElimination() {
        let currentSolved = this.nSolved
        for (let i = 0; i < this.size; i++){
            let row = []
            for (let j = 0; j < this.size; j++){
                row.push([i,j])
            }
            await this.solveGroup(row)
        }

        for (let j = 0; j < this.size; j++){
            let col = []
            for (let i = 0; i < this.size; i++){
                col.push([i,j])
            }
            await this.solveGroup(col)
        }

        for (let x = 0; x < this.nSecsH; x++){
            for (let y = 0; y < this.nSecsW; y++){
                let tlX = this.secH * x
                let tlY = this.secW * y
                let sec = []
                for (let i = 0; i < this.secH; i++){
                    for (let j = 0; j < this.secW; j++){
                        sec.push([tlX+i,tlY+j])
                    }
                }
                await this.solveGroup(sec)
            }
        }

        return (this.nSolved > currentSolved)
    }

    async solveGroup(group){
        let numbers = this.allNumbers()
        let emptys = []
        for (let pos of group){
            if (this.solved[pos[0]][pos[1]]){
                let index = numbers.indexOf(this.possible[pos[0]][pos[1]][0])
                if (index != -1){
                    numbers.splice(index,1)
                } else {
                    this.status = Status.IMPOSSIBLE
                    return false
                }
            } else {
                emptys.push(pos)
            }
        }

        for (let n of numbers){
            let chosen = null
            let found = false
            for (let pos of emptys){
                if (this.solved[pos[0]][pos[1]]){
                    continue
                }
                if (this.possible[pos[0]][pos[1]].includes(n)){
                    if (found){
                        found = false
                        break
                    } else {
                        chosen = pos
                        found = true
                    }
                }
            }
            if (found){
                await this.writeIn(n,chosen[0],chosen[1])
                await this.doQueue()
            }
        }

        return true
    }

    async doQueue() {
        let currentSolved = this.nSolved
        while (this.queue.length != 0){
            let curr = this.queue[0]
            this.writeIn(curr[0],curr[1],curr[2])
            this.queue.shift()
        }
        return (this.nSolved > currentSolved)
    }

    clone() {
        let newGrid = new Grid(this.size,this.secW,this.secH,this.depth+1)
        for (let i = 0; i < this.size; i++){
            for (let j = 0; j < this.size; j++){
                if (this.solved[i][j]){
                    newGrid.writeIn(this.possible[i][j][0],i,j)
                }
            }
        }
        return newGrid
    }

    getNatXY(x,y){
        if (this.solved[x][y]){
            return this.possible[x][y][0]
        } else {
            return 0
        }
    }

    getStatus(){
        return self.status
    }

    adapt(otherGrid){
        for (let i = 0; i < this.size; i++){
            for (let j = 0; j < this.size; j++){
                this.writeIn(otherGrid.getNatXY(i,j),i,j)
            }
        }
        this.status = otherGrid.getStatus()
    }

    async solveByGuess(){
        if (this.status != Status.IN_PROGRESS){
            return false
        }

        let changed = false
        for (let i = 0; i < this.size; i++){
            for (let j = 0; j < this.size; j++){
                if (!this.solved[i][j] && this.possible[i][j].length <= 4){
                    let works = -1
                    let nWorks = 0
                    // let attempts = []
                    for (let n of this.possible[i][j]) {
                        let tempGrid = this.clone()
                        tempGrid.writeIn(n,i,j)
                        // attempts.push(tempGrid)
                        // console.log(["Guessing",this.depth,n,i,j])
                        let status = await tempGrid.solve()
                        // console.log(JSON.parse(JSON.stringify(status)));
                        if (status === Status.SOLVED) {
                            // console.log("FOUND ONE THAT WORKS")
                            nWorks = 1
                            this.adapt(tempGrid)
                            this.status = Status.SOLVED
                            return true
                        } else if (status != Status.IMPOSSIBLE){
                            nWorks++
                            works = n
                            if (nWorks >= 2){
                                break
                            }
                        }
                    }

                    if (nWorks == 1){
                        changed = true
                        this.writeIn(works,i,j)
                        let progress = true
                        while (progress){
                            progress = await this.doQueue() || await this.solveByElimination()
                        }
                    }
                }
            }
        }

        return changed
    }

    async solve() {
        let progress = true
        while (progress){
            // console.log("going")
            progress = await this.doQueue() || await this.solveByElimination()
        }

        if (this.status == Status.IN_PROGRESS){
            // console.log("Done all logical steps")
            if (this.depth < 3){
                let dummy = await this.solveByGuess()
            }
        }

        return this.status
    }

    getGrid() {
        let ret = []
        for (let i = 0; i < this.size; i++){
            let row = []
            for (let j = 0; j < this.size; j++){
                if (this.solved[i][j]){
                    row.push(this.possible[i][j][0])
                } else {
                    row.push(0)
                }
            }
            ret.push(row)
        }
        return ret
    }
}

async function main(){
    let evilNumbers = [
        [0,0,0,5,0,8,0,2,0],
        [0,8,0,6,0,0,0,0,9],
        [0,3,5,4,0,0,7,0,0],
        [0,9,0,7,1,4,0,0,5],
        [0,0,0,0,0,0,9,0,0],
        [0,0,8,0,0,0,0,0,0],
        [0,0,0,0,0,0,1,5,0],
        [0,5,4,0,0,0,0,0,0],
        [6,0,3,0,0,2,0,0,8]
    ]
    let specNumbers = [
        [0,0,0,4,0,0,0,0,2],
        [0,0,3,0,5,0,4,0,0],
        [0,2,0,6,0,0,0,9,0],
        [1,0,7,0,0,0,0,0,0],
        [0,8,0,0,0,0,0,2,0],
        [0,0,0,0,0,0,9,0,5],
        [0,3,0,0,0,8,0,1,0],
        [0,0,2,0,1,0,5,0,0],
        [4,0,0,0,0,3,0,0,0]
    ]
    let easyNumbers = [
        [8,0,0,3,7,0,0,6,0],
        [0,9,4,0,2,0,0,0,0],
        [0,7,3,0,8,6,0,0,0],
        [0,0,0,8,1,4,0,0,5],
        [4,2,0,9,0,0,1,0,0],
        [7,1,8,0,0,2,3,0,4],
        [0,4,0,7,9,0,0,2,8],
        [5,0,2,1,0,0,0,0,0],
        [0,8,0,0,4,5,0,1,0]
    ]
    let ctcNumbers = [
        [1,0,0,4,0,0,7,0,0],
        [0,2,0,0,5,0,0,8,0],
        [0,0,3,0,0,6,0,0,9],
        [0,1,0,0,4,0,0,7,0],
        [0,0,2,0,0,5,0,0,8],
        [9,0,0,3,0,0,6,0,0],
        [7,0,0,0,0,8,0,0,2],
        [8,0,0,2,0,0,9,0,0],
        [0,9,0,0,7,0,0,1,0]
    ]
    let ctc2Numbers = [
        [0,0,0,0,0,0,0,1,2],
        [0,0,0,0,0,0,3,4,5],
        [0,0,0,0,0,3,6,7,0],
        [0,0,0,0,8,1,5,0,0],
        [0,0,0,7,5,4,0,0,0],
        [0,0,4,2,3,0,0,0,0],
        [0,6,7,9,0,0,0,0,0],
        [3,1,2,0,0,0,0,0,0],
        [8,5,0,0,0,0,0,0,0]
    ]
    let exp2Numbers = [
        [8,0,0,0,0,0,0,0,0],
        [0,0,3,6,0,0,0,0,0],
        [0,7,0,0,9,0,2,0,0],
        [0,5,0,0,0,7,0,0,0],
        [0,0,0,0,4,5,7,0,0],
        [0,0,0,1,0,0,0,3,0],
        [0,0,1,0,0,0,0,6,8],
        [0,0,8,5,0,0,0,1,0],
        [0,9,0,0,0,0,4,0,0],
    ]
    let worstNumbers = [
        [0,0,0,8,0,1,0,0,0],
        [0,0,0,0,0,0,4,3,0],
        [5,0,0,0,0,0,0,0,0],
        [0,0,0,0,7,0,8,0,0],
        [0,0,0,0,0,0,1,0,0],
        [0,2,0,0,3,0,0,0,0],
        [6,0,0,0,0,0,0,7,5],
        [0,0,3,4,0,0,0,0,0],
        [0,0,0,2,0,0,6,0,0]
    ]
    let mainGrid = new Grid(9,3,3)
    await mainGrid.load(ctc2Numbers)
    var t0 = performance.now()
    let result = await mainGrid.solve()
    var t1 = performance.now()
    // console.log("Took " + (t1 - t0) + " milliseconds.")
    // console.log(mainGrid.getGrid())
    // console.log(result)
    // return mainGrid.getGrid()
}